validator/proposer/service.go

```diff
@@ -68,10 +84,72 @@ func (p *Proposer) Stop() error {
 	return nil
 }
 
+// DoesAttestationExist checks if an attester has already attested to a block.
+func (p *Proposer) DoesAttestationExist(attestation *pbp2p.AttestationRecord) bool {
+	exists := false
+	for _, record := range p.pendingAttestation {
+		if bytes.Equal(record.GetAttesterBitfield(), attestation.GetAttesterBitfield()) {
+			exists = true
+			break
+		}
+	}
+	return exists
+}
+
+// Adds a pending attestation to the memory so that it can be included in the next
+// proposed block.
+func (p *Proposer) AddPendingAttestation(attestation *pbp2p.AttestationRecord) {
+	p.pendingAttestation = append(p.pendingAttestation, attestation)
+}
+
+// AggregateAllSignatures aggregates all the signatures of the attesters. This is currently a
+// stub for now till BLS/other singature schemes are implemented.
+func (p *Proposer) AggregateAllSignatures(attestations []*pbp2p.AttestationRecord) []uint32 {
+	// TODO: Implement Signature Aggregation.
+	return []uint32{}
+}
+
+// GenerateBitmask creates the attestation bitmask from all the attester bitfields in the
+// attestation records.
+func (p *Proposer) GenerateBitmask(attestations []*pbp2p.AttestationRecord) []byte {
+	// TODO: Implement bitmask where all attesters bitfields are aggregated.
+	return []byte{}
+}
+
+// processAttestation processes incoming broadcasted attestations from the beacon node.
+func (p *Proposer) processAttestation(done <-chan struct{}) {
+	attestationSub := p.p2p.Subscribe(&pbp2p.AttestationBroadcast{}, p.attestationBuf)
+	defer attestationSub.Unsubscribe()
+
+	for {
+		select {
+		case <-done:
+			log.Debug("Proposer context closed, exiting goroutine")
+			return
+		case msg := <-p.attestationBuf:
+			data, ok := msg.Data.(*pbp2p.AttestationBroadcast)
+			if !ok {
+				log.Error("Received malformed attestation p2p message")
+				continue
+			}
+
+			attestationRecord := data.GetAttestationRecord()
```

Comment:
Can you explain briefly what attestationRecord and attestationExists are used for?

Reply:
attestationRecord is just another name for the attestation by an attester. Attestation exists is to check that there is no duplicated attestation added to the pending attestations by the same attester

Reply:
Got it. Thanks!